### ISSUE TITLE
Hints on the framework design and entry points for integrating the motor current control in the motor parameters estimation

### ISSUE THREAD

**Issue by nunoguedelha** [robotology-playground/sensors-calib-inertial#19]

**Comment by nunoguedelha** [comment]
CC @traversaro @DanielePucci

**Comment by traversaro** [comment]
> @prashanthr05, in case the motor velocity measurement is not provided by the hardware, you should stub the respective method getMotorEncoderSpeeds in the remapper, with a code that estimates the velocity from other measurements (also considering the gearbox ratio). 

I am not sure about this. If no low-level speed estimates are available, the proper replacement is a non-causal velocity estimation, that cannot be done in realtime without delay that could affect the estimation.

**Comment by nunoguedelha** [comment]
@traversaro , yes indeed. Sorry for the confusion, I was focusing on the interface for getting the velocity. The idea would be to have a stub that gets the last results from a velocity estimator. That estimator could eventually even be implemented in C++ and running under a different thread.

**Comment by traversaro** [comment]
Perhaps we need to discuss this f2f, but what I meant is that for estimation we should use a non-causal velocity estimation, that could be implemented in matlab.

**Comment by nunoguedelha** [comment]
ok, I get it. Anyway we can discuss further after lunch.

**Comment by prashanthr05** [comment]
> I am not sure about this. If no low-level speed estimates are available, the proper replacement is a non-causal velocity estimation, that cannot be done in realtime without delay that could affect the estimation.

@traversaro To have clarity,  in our case (if I'm not wrong), the joint velocities are obtained as a result of the velocity estimator, however we do not have a measure or an estimate for the motor velocities. But this could be obtained by mapping the joint velocities through the transmission. But, it would be better to not rely on these velocity estimates due to the inherent delay involved in the controller.

Hence we would rely on offline non-causal estimation of the velocities and accelerations by applying a savitzy-golay filtering over the joint positions.  Correct me, if I'm wrong ?

**Comment by prashanthr05** [comment]
> ok, I get it. Anyway we can discuss further after lunch.

I'm sorry, I'm not in IIT. I could join you through skype, if you're going to have a meeting/discussion on this.

**Comment by lrapetti** [comment]
Before using current control, I require to do current control calibration similarly to `LowlevTauCtrlCalibrator`, and I am going to implement a [`LowlevCurrCtrlCalibrator`](https://github.com/lrapetti/sensors-calib-inertial/tree/task/LowlevCurrCtrlCalibrator) task (to compute PWM-current constant and BEMF friction constant). 

In order to do that I need to record current measurement that is currently not available in `sensor-calib-intertial`.

I have some questions regarding this implementation. @nunoguedelha

* The current, as the other quantities, is written on the `StateExt` port. However it is not parsed in [readStateExt.m](https://github.com/robotology-playground/sensors-calib-inertial/blob/master/external/yarp/readStateExt.m), and the `getCurrents` method is not defined in the `sensor-calib-intertial` [ControlBoardRemapper](https://github.com/robotology-playground/sensors-calib-inertial/tree/master/src/%40RemoteControlBoardRemapper).
I think I can work on adding current measurement, but I am probably missing some passages of the sensors measurement process. It would help if you could give me some extra info on sensors measurement framework in `sensor-calib-intertial`?

* When a profile is defined, clearly the option `['meas' 'joint']` enable the measurement of `qs, dqs, d2qs` (joint position, velocity and acceleration) while `['meas' 'jtorq']` enables the measurement of `taus` (torques). I noticed there is no specific option for `dqMs` (motor velocity) and `pwms`, that anyway seems to be enabled by the joints measurement option `['meas' 'joint']`. 
Did I get it right? 
Considering that I want to measure the current, does it need to be enabled by a different option ['meas' 'curr'], or should it be included in `['meas' 'joint']` option?

**Comment by nunoguedelha** [comment]
### Enabling measurements for specific sensors
<br/>
<img src="https://user-images.githubusercontent.com/6848872/42643311-1f8f11f4-85f9-11e8-809f-136ac1eb06fe.png" width="250" align="left">

> "It would help if you could give me some extra info on sensors measurement framework in sensor-calib-intertial?"

I forgot to mention in the initial notes... The compomemt `SensorDataYarpI` (snapshot on the left) handles the ports required for acquiring data from specific sensors. This is defined jointly by the config file https://github.com/robotology-playground/sensors-calib-inertial/blob/master/src/conf/yarpPortNameRules.m and the method `buildSensType2portNgetter(...)`:
https://github.com/robotology-playground/sensors-calib-inertial/blob/dacb506e6dc92bde46eacf6211a78a517d3cca95/src/%40SensorDataYarpI/SensorDataYarpI.m#L123-L147

This function maps the sensor types to the Yarp ports where the respective data can be retrieved. The entries to that map are the sensor types you find in the Motion Sequence Profile files like https://github.com/robotology-playground/sensors-calib-inertial/blob/master/src/conf/advanced/lowLevTauCtrlCalibratorSequenceProfile1.m. This way `['joint' , 'from']` will map to the `/icub/<part>/stateExt:o` port.

The initial idea was to have an interface of the profile files independent w.r.t. how the ports group the sensor data on iCub, so although there are two labels 'joint' (for joint encoders) and 'jtorq', they both map to the same port on iCub. For some reason I forgot (probably lack of time), I didn't define a label for 'pwm' ad 'mot', but just reused 'joint'. So the option `{'meas','joint'}` is the one used.
For that to work, in the `stateExt:o` port, we parse all the available measurements **joint q, dq, d2q, motor dq, and pwm**.

> Considering that I want to measure the current, does it need to be enabled by a different option ['meas' 'curr'], or should it be included in ['meas' 'joint'] option?

As you like, I suggest you use ['meas' 'curr'] and on the go, correct the case for 'pwm' and 'mot' (motor q,dq,d2q).

### Retrieving the data offline from the port `readStateExt.m`

I let you take a look at the script and we can rediscuss tomorrow.

**Comment by nunoguedelha** [comment]
@lrapetti , could you move this to a new issue please? in order to leave in this one only the tutorial discussion.
